File: swift/09_queue/QueueBasedOnLinkedList.swift
//
// Created by Jiandan on 2018/10/11.
// Copyright (c) 2018 Jiandan. All rights reserved.
//

import Foundation

struct QueueBasedOnLinkedList<Element>: Queue {
    /// 队首
    var head: Node<Element>?
    /// 队尾
    var tail: Node<Element>?
    
    // MARK: Protocol: Queue
    
    var isEmpty: Bool { return head == nil }
    
    var size: Int {
        if isEmpty {
            return 0
        }
        
        var count = 1 // head 本身算一个
        while head?.next != nil {
            count += 1
        }
        
        return count
    }
    
    var peek: Element? { return head?.value }
    
    mutating func enqueue(newElement: Element) -> Bool {
        if isEmpty {
            // 空队列
            let node = Node(value: newElement)
            head = node
            tail = node
        } else {
            tail!.next = Node(value: newElement)
            tail = tail!.next
        }
        return true
    }
    
    mutating func dequeue() -> Element? {
        if isEmpty {
            return nil
        }
        
        let node = head
        head = head!.next
        return node?.value
    }
}

File: javascript/09_queue/QueueBasedOnLinkedList.js
/**
 * 基于链表实现的队列。
 *
 * Author: nameczz
 */

class Node {
    constructor(element) {
        this.element = element
        this.next = null
    }
}

class QueueBasedOnLinkedList {
    constructor() {
        this.head = null
        this.tail = null
    }

    enqueue(value) {
        if (this.head === null) {
            this.head = new Node(value)
            this.tail = this.head
        } else {
            this.tail.next = new Node(value)
            this.tail = this.tail.next
        }
    }

    dequeue() {
        if (this.head !== null) {
            const value = this.head.element
            this.head = this.head.next
            return value
        } else {
            return -1
        }
    }
}
// Test
const newQueue = new QueueBasedOnLinkedList()
// 插入元素
newQueue.enqueue(1)
newQueue.enqueue(2)
newQueue.enqueue(3)
// 获取元素
let res = 0
console.log('-------获取dequeue元素------')
while (res !== -1) {
    res = newQueue.dequeue()
    console.log(res)
}
